File: BruteForce.py
from Helper import read_input

class BruteForceSolver():

    def __init__(self, n_items, n_trucks, items, trucks):
        self.n_items = n_items
        self.n_trucks = n_trucks
        self.items = items
        self.trucks = trucks

        self.f = 0 
        self.f_best = 1e9

        self.t = [-1]*self.n_items
        self.x = [-1]*self.n_items
        self.y = [-1]*self.n_items
        self.o = [-1]*self.n_items
        self.t_solution = [-1]*self.n_items
        self.x_solution = [-1]*self.n_items
        self.y_solution = [-1]*self.n_items
        self.o_solution = [-1]*self.n_items

        self.prior_trucks = list(range(self.n_trucks))
        self.prior_trucks.sort(key=lambda k: self.trucks[k][0] * self.trucks[k][1], reverse=True)
    
    def printSolution(self, opt=False):
        if opt:
            for i in range(self.n_items):
                print(i+1, self.t_solution[i]+1, self.x_solution[i], self.y_solution[i], self.o_solution[i])
        else:
            for i in range(self.n_items):
                print(i+1, self.t[i]+1, self.x[i], self.y[i], self.o[i])   

    def isOverlapping(self, x1, y1, w1, h1, x2, y2, w2, h2):
        return max(0,min(x1 + w1, x2 + w2) - max(x1, x2)) * max(0, min(y1 + h1, y2 + h2) - max(y1, y2)) > 0

    def check(self, n, k, xn, yn, on):
        w, l = self.items[n][0], self.items[n][1]
        W, L = self.trucks[k][0], self.trucks[k][1]
        if(on==1):
            w, l = l, w

        if((0<=xn) and (xn+w <= W) and (0<=yn) and (yn+l <= L)):
            for i in range(n):
                if(self.t[i] == k):
                    wi, li = self.items[i][0], self.items[i][1]
                    xi, yi = self.x[i], self.y[i]
                    if(self.o[i] == 1):
                        wi, li = li, wi

                    if(self.isOverlapping(xi, yi, wi, li, xn, yn, w, l)):
                        return False
            return True
        return False
    
    def update(self, n, k, xn, yn, on):
        self.t[n] = k
        self.x[n] = xn
        self.y[n] = yn
        self.o[n] = on

    def restore(self, n):
        self.t[n] = -1
        self.x[n] = -1
        self.y[n] = -1
        self.o[n] = -1

    def calCurrentCost(self, n):
        f=0
        inTruck = [0]*self.n_trucks
        for i in range(n):
            inTruck[self.t[i]] += 1
        for i in range(self.n_trucks):
            if(inTruck[i] > 0):
                f += self.trucks[i][2]
        return f
    
    def solve(self, n):
        for tr in range(self.n_trucks):
            k = self.prior_trucks[tr]; 
            m = min(self.items[n][0], self.items[n][1])
            for xn in range(self.trucks[k][0] - m):
                for yn in range(self.trucks[k][1] - m):
                    for on in range(2):
                        if(self.check(n, k, xn, yn, on)):
                            self.update(n, k, xn, yn, on)
                            if(n == self.n_items-1):
                                f = self.calCurrentCost(n)
                                if(f < self.f_best):
                                    self.f_best = f
                                    for i in range(self.n_items):
                                        self.t_solution[i] = self.t[i]
                                        self.x_solution[i] = self.x[i]
                                        self.y_solution[i] = self.y[i]
                                        self.o_solution[i] = self.o[i]
                            else:
                                f = self.calCurrentCost(n)
                                if(f<self.f_best):
                                    self.solve(n+1)
                            self.restore(n)
    
    def writeToFile(self, file_path):
        with open(file_path, "w") as f:
            for i in range(self.n_items):
                f.writelines(f"{i+1} {self.t_solution[i]+1} {self.x_solution[i]} {self.y_solution[i]} {self.o_solution[i]}\n")

if __name__ == "__main__":
    input_path = "E:/planning_optimization/container2Dloading/data.in"
    output_path = "E:/planning_optimization/container2Dloading/data.out"

    n_items, n_trucks, items, trucks = read_input(file_path=input_path)

    solver = BruteForceSolver(n_items, n_trucks, items, trucks)
    solver.solve(0)
    solver.printSolution(opt=True)
    solver.writeToFile(output_path)

File: BruteForce.cpp
#include <bits/stdc++.h>
#define NMAX 1000
#define KMAX 1000

using namespace std;

int n_items, n_trucks;
int items[NMAX][2]; //items[i][0] is w[i], items[i][1] is l[i]
int trucks[KMAX][3]; //trucks[k][0] is W[k], trucks[k][1] is L[k], trucks[k][2] is c[k]
int priorTrucks[KMAX];
int t[NMAX], x[NMAX], y[NMAX], o[NMAX];
int t_solution[NMAX], x_solution[NMAX], y_solution[NMAX], o_solution[NMAX];
int f=0, f_best=INT_MAX;

bool compare(int &a, int &b) {
    int value_a = trucks[a][0]*trucks[a][1];
    int value_b = trucks[b][0]*trucks[b][1];
    return value_a > value_b;
}

void sortTrucks(){
    for(int k=0; k<n_trucks; ++k){
        priorTrucks[k] = k;
    }
    sort(priorTrucks, priorTrucks + n_trucks, compare);
}

void init(){
    for(int i=0; i<n_items; ++i){
        t[i] = -1; t_solution[i] = -1;
        x[i] = -1; x_solution[i] = -1;
        y[i] = -1; y_solution[i] = -1;
        o[i] = -1; o_solution[i] = -1;
    }

    sortTrucks();
}

void printOptSolution(){
    for(int i=0; i<n_items; ++i){
        cout<<i+1<<" "<<t_solution[i]+1<<" "<<x_solution[i]<<" "<<y_solution[i]<<" "<<o_solution[i]<<"\n";
    }
}

void printSolution(){
    for(int i=0; i<n_items; ++i){
        cout<<i+1<<" "<<t[i]+1<<" "<<x[i]<<" "<<y[i]<<" "<<o[i]<<"\n";
    }    
}

int isOverlapping(int x1, int y1, int w1, int h1, int x2, int y2, int w2, int h2) {
    if(x1 + w1 <= x2) return false;
    if(x2 + w2 <= x1) return false;
    if(y1 + h1 <= y2) return false;
    if(y2 + h2 <= y1) return false;
    return true;
}

bool check(int n, int k, int xn, int yn, int on){
    int w=items[n][0], l=items[n][1];
    int W=trucks[k][0], L=trucks[k][1];
    if(on==1){
        swap(w,l);
    }
    if((0<=xn) && (xn+w <= W) && (0<=yn) && (yn+l <= L)){
        for(int i=0; i<n; ++i){
            if(t[i] == k){
                int wi = items[i][0], li = items[i][1];
                int xi = x[i], yi = y[i];
                if(o[i] == 1){
                    swap(wi, li);
                }
                if(isOverlapping(xi, yi, wi, li, xn, yn, w, l)){
                    return false;
                }
            }
        }
        return true;
    }
    return false;
}

void update(int n, int k, int xn, int yn, int on){
    t[n] = k;
    x[n] = xn;
    y[n] = yn;
    o[n] = on;
}

void restore(int n, int k, int xn, int yn, int on){
    t[n] = -1;
    x[n] = 0;
    y[n] = 0;
    o[n] = 0;
}

int calCurrentCost(int n){
    int f=0;
    int inTruck[KMAX]={0};
    for(int i=0; i<n; ++i){
        inTruck[t[i]] += 1;
    }
    for(int i=0; i<n_trucks; ++i){
        if(inTruck[i] > 0){
            f += trucks[i][2];
        }
    }   
    return f;
}

void Try(int n){
    for(int tr=0; tr<n_trucks; ++tr){
        int k = priorTrucks[tr]; 
        int m = min(items[n][0],items[n][1]);
        for(int xn=0; xn<trucks[k][0] - m; xn+=10){
            for(int yn=0; yn<trucks[k][1] - m; yn+=10){
                for(int on=0; on<2; ++on){
                    if(check(n, k, xn, yn, on)){
                        update(n, k, xn, yn, on);
                        if(n == n_items-1){
                            int f = calCurrentCost(n);
                            if(f < f_best){
                                f_best = f;
                                for(int i=0; i<n_items; ++i){
                                    t_solution[i] = t[i];
                                    x_solution[i] = x[i];
                                    y_solution[i] = y[i];
                                    o_solution[i] = o[i];
                                }
                            }
                        }else{
                            int f = calCurrentCost(n);
                            if(f<f_best){
                                Try(n+1);
                            }
                        }
                        restore(n, k, xn, yn, on);
                    }               
                }
            }
        }
    }
}

int main(){
    freopen("INPUT.txt", "r", stdin);
    // freopen("OUTPUT.txt", "w", stdout);
    cin>>n_items>>n_trucks;
    for(int i=0; i<n_items; ++i){
        cin>>items[i][0]>>items[i][1];
    }
    for(int k=0; k<n_trucks; ++k){
        cin>>trucks[k][0]>>trucks[k][1]>>trucks[k][2];
    }
    init();
    Try(0);
    printOptSolution();
    return 0;
}